### PR TITLE
Add support for fork()

### PR DESCRIPTION
**Pull request by cstanfill**
- Keep track of what pages users allocate in a pagemap struct
- On a fork(), copy all mapped pages while interpreting the syscall in the parent.
- We don't actually remap the pages (in the child's pagetable) until it gets scheduled (code runs in finish_fork(0)).
- test-memory now tests that fork() behaves as expected
- the pagefault handler now prints the address that faulted and the thread id.
- I'm definitely not sure if the fpu state is properly carried over

### CODE REVIEW

**Comment by cstanfill** [comment]
This fixes #14 as well.

File: userland/test-memory/expected.txt [hunk 1]
```diff
@@ -3,3 +3,4 @@ THREAD 0x01: Reading
 THREAD 0x01: Freeing
 THREAD 0x01: Reading
 THREAD 0x01: Page faulting
+THREAD 0x01: user pf: 0x0000000000004000
```
Comment (gsilvis) [inline userland/test-memory/expected.txt]: Ideally, this message would look different from the messages from userspace.

File: kernel/pagemap.c [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  if (phys && !ret) {
+    *phys = ent.phys;
+  }
+
+  return ret;
+}
+
+int pm_is_mapped (pagemap *p, uint64_t virt) {
+  return pm_get_ent(p, virt, 0) == 0;
+}
```
Comment (gsilvis) [inline kernel/pagemap.c]: Some of these take a const pagemap*

File: kernel/page.c [hunk 3]
```diff
@@ -113,7 +113,26 @@ int map_new_page (uint64_t virt, uint64_t mode) {
...
+    dst->entries[i].phys = virt_to_phys((uint64_t) new_page) | (src->entries[i].phys & PAGE_4K_MASK);
+  }
+  dst->num_entries = src->num_entries;
+}
+
+void apply_pagemap (void) {
+  pagemap *pm = &running_tcb->pm;
+  for (uint8_t i = 0; i < pm->num_entries; ++i) {
+    pagemap_ent ent = pm->entries[i];
+    insert_page(PAGE_4K_ALIGN(ent.phys), ent.virt, ent.phys & PAGE_4K_MASK);
```
Comment (gsilvis) [inline kernel/page.c]: careful of NX

File: kernel/threads.c [hunk 4]
```diff
@@ -146,3 +146,42 @@ void block_current_thread (void) {
...
+      /* Set up stacks */
+      char *kernel_stack = &((char *)allocate_phys_page())[4096];
+      /* Initialize tcb struct */
+      tcbs[i].state = TS_INACTIVE;
+      tcbs[i].stack_top = &kernel_stack[0];
+      memcpy(kernel_stack - stack_depth,
+          ((char *)running_tcb->stack_top) - stack_depth, stack_depth);
+      /* Initialize stack */
+      tcbs[i].rsp = kernel_stack - stack_depth;
+      tcbs[i].fpu_state = THREAD_FPU_STATE_INACTIVE;
```
Comment (gsilvis) [inline kernel/threads.c]: Set it to the fpu_state of the old thread; allocate and copy the fpu_buf if it's ACTIVE

File: kernel/page.c (hunk 3 above)
Comment (cstanfill) [inline kernel/page.c]: NX is stupid

File: kernel/syscall-defs.h [hunk 5]
```diff
@@ -13,7 +13,8 @@ typedef unsigned long long syscall_arg;
 #define SYSCALL_PFREE       0x07
 #define SYSCALL_DEBUG       0x08
 #define SYSCALL_NANOSLEEP   0x09
+#define SYSCALL_FORK        0x0A
```
Comment (allanlw) [inline kernel/syscall-defs.h]: Wow! it's so easy to add a syscall now!!

Comment (gsilvis) [inline kernel/syscall-defs.h]: It'd be even easier if it were an enum :)

File: kernel/threads-asm.s [hunk 6]
```diff
@@ -1,3 +1,21 @@
+.macro push_callee_save_reg
```
Comment (allanlw) [inline kernel/threads-asm.s]: Aren't these also defined in isr-asm.s ? Can they be shared?

File: kernel/threads.c [hunk 7]
```diff
@@ -143,3 +146,42 @@ void block_current_thread (void) {
...
   schedule();
 }
+
+int fork_ret = 0;
+
+int clone_thread (uint64_t fork_rsp) {
+  /* This tells us how much of the kernel stack to copy to the new thread */
+  uint64_t stack_depth = (uint64_t) running_tcb->stack_top - fork_rsp;
+  for (int i = 0; i < NUMTHREADS; i++) {
+    if (tcbs[i].state == TS_NONEXIST) {
```
Comment (allanlw) [inline kernel/threads.c]: A nice-to-have would probably be having a function that does get_free_tcb or something so this loop doesn't get written twice.

File: kernel/threads.c (hunk 4 above)
Comment (cstanfill) [inline kernel/threads.c]: Also refactoring this to reuse more code from user_thread_create.

File: kernel/threads-asm.s (hunk 6 above)
Comment (cstanfill) [inline kernel/threads-asm.s]: isr-asm.s defines push_call**er**_save_reg, as opposed to call**ee**. So no.

File: kernel/threads.c (hunk 7 above)
Comment (cstanfill) [inline kernel/threads.c]: George mentioned this as a nice-to-have to me IRL as well, so I got around to doing it.

**Comment by allanlw** [comment]
A higher-level design question is if the fork API should be the same as on unix-like OSes. Reasonably, fork could take some flags for controlling how to fork or could return more information in a struct via a pointer out arg.

For example, consider a fork_info struct, containing information like the parent's pid, the child's pid, etc. This model could also allow for octopus (non-binary) forking, as then you're not restricted to one return pid.

See also: linux's `clone`

File: kernel/threads.c [hunk 8]
```diff
@@ -16,40 +16,51 @@
 
 tcb tcbs[NUMTHREADS];
 
-
-/* Returns 0 on success, negative on failure */
-int user_thread_create (void *text, size_t length) {
+/* After calling this, you must set up the kernel stack contents, rsp, and fpu
+ * state if not INACTIVE. */
+int create_thread (void* text, size_t length, tcb** out) {
```
Comment (gsilvis) [inline kernel/threads.c]: I'd prefer  ``tcb *create_thread (void* text, size_t length)``, returning NULL if there's no more room

**Comment by gsilvis** [comment]
That sounds like good future work, IMO.  For now, fork is so helpful for testing that I'm fine with a suboptimal/boring interface

File: kernel/threads-asm.s (hunk 6 above)
Comment (allanlw) [inline kernel/threads-asm.s]: That's stupid. They should be defined in the same place.

File: kernel/threads.c (hunk 8 above)
Comment (cstanfill) [inline kernel/threads.c]: Plebeians!!!

File: userland/test-memory/main.c [hunk 9]
```diff
@@ -5,41 +5,55 @@ inline unsigned char* getpage(unsigned int i) {
...
-    // Use unsigned char to prevent problems when extending to unsigned longs.
     unsigned char* my_page = getpage(i);
     int ret = palloc(my_page);
     if (ret) {
       DEBUG("Failed to allocate page!");
       exit();
     }
+  }
+
+  DEBUG("Forking");
```
Comment (allanlw) [inline userland/test-memory/main.c]: This program could more robustly test the fork syscall by testing the cases of:

- Making sure the kernel allows more than one fork, ever.
- Making sure the kernel allows both the parent and child to fork correctly after forking once.
- Making sure the kernel allows more than one child fork (bottom half) to be pending at once.

File: kernel/threads.c [hunk 10]
```diff
@@ -143,3 +155,34 @@ void block_current_thread (void) {
   running_tcb->state = TS_BLOCKED;
   schedule();
 }
+
+int fork_ret = 0;
```
Comment (allanlw) [inline kernel/threads.c]: Seeing as this is only used in threads-asm.s it should probably also be defined there, without global scoping.

File: userland/test-memory/main.c (hunk 9 above)
Comment (cstanfill) [inline userland/test-memory/main.c]: Two threads ought to be enough for anybody. I will add an octopus test tomorrow though.

File: kernel/page.c [hunk 11]
```diff
@@ -64,33 +60,80 @@ int map_page (uint64_t phys, uint64_t virt, uint64_t mode) {
...
+  }
+  uint64_t phys = virt_to_phys((uint64_t)allocate_phys_page());
+  /* We checked that the pagemap wasn't full before, so this can't fail. */
+  pm_add_ent(pm, virt, phys | mode);
+  insert_page(phys, virt, mode);
+  return 0;
+}
+
+void clone_pagemap (pagemap *dst, pagemap *src) {
+  /* TODO: Copy-on-write. */
```
Comment (allanlw) [inline kernel/page.c]: COWs go MOO

Comment (gsilvis) [inline kernel/page.c]: fish go blub!

**Pull request merged**